vcd/resource_vcd_cloned_vapp.go

```diff
@@ -0,0 +1,247 @@
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "vApp Hyper Reference",
+			},
+		},
+	}
+}
+
+func resourceVcdClonedVAppCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+	_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+	if err != nil {
+		return diag.Errorf("error retrieving Org and VDC: %s", err)
+	}
+
+	vappName := d.Get("name").(string)
+	vappDescription := d.Get("description").(string)
+	powerOn := d.Get("power_on").(bool)
+
+	sourceType := strings.ToLower(d.Get("source_type").(string))
+	sourceId := d.Get("source_id").(string)
+	deleteSource := d.Get("delete_source").(bool)
+
+	var vapp *govcd.VApp
+	if sourceType == "vapp" {
+		sourceVapp, err := vdc.GetVAppByNameOrId(sourceId, true)
+		if err != nil {
+			return diag.Errorf("error retrieving source vApp with ID '%s': %s", sourceId, err)
+		}
+		params := &types.CloneVAppParams{
+			Name:        vappName,
+			Deploy:      true,
+			PowerOn:     powerOn,
+			Description: vappDescription,
+			Source: &types.Reference{
+				HREF: sourceVapp.VApp.HREF,
+				ID:   sourceVapp.VApp.ID,
+				Type: sourceVapp.VApp.Type,
+				Name: sourceVapp.VApp.Name,
+			},
+			IsSourceDelete: addrOf(deleteSource),
+		}
+		sourceStatus, err := sourceVapp.GetStatus()
+		if err != nil {
+			return diag.Errorf("error getting the status of source vApp %s: %s", sourceVapp.VApp.Name, err)
+		}
+		if sourceStatus != "POWERED_OFF" {
+			if deleteSource {
+				return diag.Errorf("'delete_source' was requested, but the vApp '%s' is not powered off."+
+					" Either set 'delete_source' to false or power off the vApp", sourceVapp.VApp.Name)
+			}
+		}
+		vapp, err = vdc.CloneVapp(params)
+		if err != nil {
+			return diag.Errorf("error cloning vApp %s from another vApp: %s", vappName, err)
+		}
+	} else {
+		sourceTemplate, err := vcdClient.GetVAppTemplateById(sourceId)
+		if err != nil {
+			return diag.Errorf("error retrieving vApp template with ID '%s': %s", sourceId, err)
+		}
+		params := &types.InstantiateVAppTemplateParams{
+			Name:        vappName,
+			Deploy:      true,
+			PowerOn:     powerOn,
+			Description: vappDescription,
+			Source: &types.Reference{
+				HREF: sourceTemplate.VAppTemplate.HREF,
+				ID:   sourceTemplate.VAppTemplate.ID,
+				Type: sourceTemplate.VAppTemplate.Type,
+				Name: sourceTemplate.VAppTemplate.Name,
+			},
+			IsSourceDelete:   deleteSource,
+			AllEULAsAccepted: true,
+		}
+		vapp, err = vdc.CreateVappFromTemplate(params)
+		if err != nil {
+			return diag.Errorf("error creating vApp %s from template: %s", vappName, err)
+		}
+	}
+
+	d.SetId(vapp.VApp.ID)
+
+	vappStatus, err := vapp.GetStatus()
+	if err != nil {
+		return diag.Errorf("error retrieving vApp '%s' status: %s", vappName, err)
+	}
+	// If the cloned vApp is suspended and power ON was requested, we adjust the status accordingly
+	if vappStatus == "SUSPENDED" && powerOn {
+		err = vapp.DiscardSuspendedState()
```

Comment:
is there any chance that the VM is not suspended (powered off?), but power on is requested (we'd miss it in such case)?

Reply:
The PowerOn operation is handled directly by the API. The "suspended" state is a side effect that I have seen happening often, and I added an adjustment.
I assume that the API would return an error if the powering ON fails, but I couldn't manage to trigger such occurrence.